### ISSUE TITLE
Adding Build Configurations for Cross Compilation to the CI Service

### ISSUE THREAD

**Issue by lyrahgames** [build2/build2#295]
As `build2` offers first-class support for cross compilation and you already created its CI service and such a wealth of build configurations, I am interested in the question whether it would make sense to additionally provide build configurations for general cross compilation setups. As already mentioned in the build system manual, a good example would be to cross-compile Windows executables on Linux.

**Comment by boris-kolpackov** [comment]
Yes, we could probably add something like this and Linux-to-Windows with MinGW GCC shouldn't be difficult to setup. However, quite a few packages already struggle with MinGW GCC so using it to test cross-compilation might just conflate two sources of errors (package doesn't build with MinGW GCC and package doesn't build in a cross-compiling setup). Ideally we would want something that generally works, maybe `x86_64-linux` to `aarch64-linux` (or even the other way around). Not sure how difficult that will be to setup (I think Debian provides to `aarch64` cross-compiler packages). If you would like to investigate this further, that would be great.

Note also that the Emscripten configuration is cross-compiling.

**Comment by boris-kolpackov** [comment]
BTW, one thing in favor of the Linux-to-Windows setup is that we will be able to run (at least some) tests using Wine (similar to how we run them with Node.js in the Emscripten configuration). I don't think we will be able to do the same for `aarch64-linux` easily.

**Comment by lyrahgames** [comment]
> Yes, we could probably add something like this and Linux-to-Windows with MinGW GCC shouldn't be difficult to setup. However, quite a few packages already struggle with MinGW GCC so using it to test cross-compilation might just conflate two sources of errors (package doesn't build with MinGW GCC and package doesn't build in a cross-compiling setup).

I understand your point.
Would it make things better to not offer such build configs in the `default` CI class but only in a `cross` class?

> Ideally we would want something that generally works, maybe `x86_64-linux` to `aarch64-linux` (or even the other way around). Not sure how difficult that will be to setup (I think Debian provides to `aarch64` cross-compiler packages). If you would like to investigate this further, that would be great.

That sounds like an interesting idea.
So far, I have only been experimenting with cross-compilation for `riscv64` from `x86_64` on Arch Linux.
I would definitely offer to help and would also take a deeper look into this issue as soon as I will have some time to spare.

> BTW, one thing in favor of the Linux-to-Windows setup is that we will be able to run (at least some) tests using Wine (similar to how we run them with Node.js in the Emscripten configuration). I don't think we will be able to do the same for `aarch64-linux` easily.

For the `riscv64` setup, I have used `config.test.runner = qemu-riscv64 -L /usr/riscv64-linux-gnu` to run the tests with QEMU.
Maybe there is a similar method for `aarch64`. But also it was a very primitive example and I do not know what problems might occur when trying to use multiple non-trivial dependencies.

**Comment by boris-kolpackov** [comment]
> Would it make things better to not offer such build configs in the `default` CI class but only in a `cross` class?

If you don't make it part of the `default` class then it mean testing cross-compilation will be optional and that will relegate cross-compilation support to the "second-class" status. In `build2` we tried very hard to make cross-compilation support the norm rather than the exception so making cross-compilation testing optional feels like a step backwards.
